account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,406 @@
+                "Connection button Clicked - endpoint name {},".format(self.name) +
+                "redirect URL:- {}, Path url:- {}".format(self.hmrc_configuration.redirect_url, self.path)
+            )
+        elif endpoint_record.name == "mtd_hello_user_endpoint":
+            self.which_button_type_clicked = "user"
+            self.path = "/hello/user"
+            self._logger.info(
+                "Connection button Clicked - endpoint name {}, ".format(self.name) +
+                "redirect URL:- {}, Path url:- {}".format(self.hmrc_configuration.redirect_url, self.path)
+            )
+            # search for token record for the API
+            token_record = self.env['mtd.api_tokens'].search([('api_id', '=', self.api_name.id)])
+            self._logger.info(
+                "Connection button Clicked - endpoint name {} , ".format(self.name) +
+                "and the api is :- {} ".format(self.api_name)
+            )
+            if not token_record.access_token and not token_record.refresh_token:
+                self._logger.info(
+                    "Connection button Clicked - endpoint name {0}, No access token ".format(self.name) +
+                    "found and no refresh_token found from the token record table."
+                )
+                api_tracker = self.env['mtd.api_request_tracker'].search([
+                    ('request_sent', '=', True),
+                    ('response_received', '=', False)
+                ])
+                self._logger.info(
+                    "Connection button Clicked - endpoint name {}, ".format(self.name) +
+                    "Checking to see if a request is in process"
+                )
+                time_10_mins_ago = (datetime.now() - timedelta(minutes=10))
+                format_time_10_mins_ago = time_10_mins_ago.strftime('%Y-%m-%d- %H:%M:%S')
+                if format_time_10_mins_ago >= api_tracker.create_date:
+                    # this means that time is more than 10 minutes a go so we can place a new request
+                    # and close this request.
+                    api_tracker.response_received = True
+                    self._logger.info(
+                        "Connection button Clicked - endpoint name {}, ".format(self.name) +
+                        "no Pending requests"
+                    )
+                    return self.get_user_authorisation()
+                else:
+                    # THe request made was within 10 mins so the user has to wait.
+                    test = True
+                    raise exceptions.Warning(
+                        "An authorisation request is already in process!!!\n " +
+                        "Please try again later"
+                    )
+            else:
+                self._logger.info(
+                    "Connection button Clicked - endpoint name {}, ".format(self.name) +
+                    "We have access token and refresh token"
+                )
```

Comment:
I'm not sure this assertion is correct - I think `access_token` *inclusive or* `refresh_token` will be here.

## If you mean inclusive or

change the log message as appropriate

## If you need both tokens to be defined

the `if` clause should be

```
if not (token_record.access_token and token_record.refresh_token):
```

It could be worth swapping the if and else blocks so the if clause reads:

```
if token_record.access_token and token_record.refresh_token:
    ...
    return version
elif (date comparison from above):
    ...
    return self.get_user_authorisation()
else:
    raise exceptions.Warning(...)
```

Reply:
I understand what you mean but the reason I have done this if block in this way is because I don't want to always work out the date comparison only when we don't have access_tokens and refresh_token. If i do a check on the statement of true then I would always have to work out 10 mins its will add to the process time.